### ISSUE TITLE
Add loop-less disk builder (systemd-repart)

### ISSUE THREAD

**Issue by schaefi** [OSInside/kiwi#2445]
## Description

The current implementation for the disk builder makes use kernel interfaces, meaning the produced devices are created through loop devices as well as the produced file-systems gets mounted and more. Under certain conditions the dependency into the kernel could cause a problem e.g if the calling user has not enough permissions or the environment does not expose access to these interfaces. In this case it would be handy if kiwi could offer a disk builder which does not require kernel interfaces.

The systemd project offers a tool called ```systemd-repart``` which implements many parts of the build process without the need for the kernel interfaces. This leads to the idea to write a new builder class that utilizes  ```systemd-repart``` to build the disk image. systemd-repart is also used in the mkosi image builder from which the idea originates. For reference see make_image():

* https://github.com/systemd/mkosi/blob/main/mkosi/__init__.py#L2621

kiwi needs to implement an interface to ```systemd-repart``` and its required input files

**Comment by schaefi** [comment]
on suse ```systemd-repart``` is provided with the ```systemd-experimental``` package

**Comment by Conan-Kudo** [comment]
This might become more important to resolve issues around running kiwi in container environments.

**Comment by schaefi** [comment]
yes this is a known problem in containers when no udev is running inside. You can fix this as follows

Make sure your container comes with a file named `/etc/kiwi.yml` inside of the file set:

```
mapper:
  - part_mapper: kpartx
```

rebuild the image. kiwi will now use kpartx which doesn't require udev to create the device nodes.

Your container still needs to be privileged though to allow creation of device nodes

**Comment by bnicolae86** [comment]
> yes this is a known problem in containers when no udev is running inside. You can fix this as follows
> 
> Make sure your container comes with a file named `/etc/kiwi.yml` inside of the file set:
> 
> ```
> mapper:
>   - part_mapper: kpartx
> ```
> 
> rebuild the image. kiwi will now use kpartx which doesn't require udev to create the device nodes.
> 
> Your container still needs to be privileged though to allow creation of device nodes

thanks, I'll try it

**Comment by bnicolae86** [comment]
> > yes this is a known problem in containers when no udev is running inside. You can fix this as follows
> > Make sure your container comes with a file named `/etc/kiwi.yml` inside of the file set:
> > ```
> > mapper:
> >   - part_mapper: kpartx
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > rebuild the image. kiwi will now use kpartx which doesn't require udev to create the device nodes.
> > Your container still needs to be privileged though to allow creation of device nodes
> 
> thanks, I'll try it

@schaefi it worked but now I have another issue trying to use gitlab.com runners. it seems that they don't have support for BTRFS filesystems :(

**Comment by aplanas** [comment]
That is a cool idea!  For now `systemd-repart` cannot be used without root if btrfs subvolumes are used. Is there a workaround than kiwi can use?

**Comment by schaefi** [comment]
> That is a cool idea! For now `systemd-repart` cannot be used without root if btrfs subvolumes are used. Is there a workaround than kiwi can use?

The only workaround I know is building in VM as standard user that has to have access to `/dev/kvm` or to build with noaccel (very slow). We call this `boxbuild`. More information here:

* https://osinside.github.io/kiwi/plugins/self_contained.html

**Comment by DaanDeMeyer** [comment]
@aplanas Latest btrfs-progs can create subvolumes without root and systemd-repart has been updated to make use of this.

**Comment by aplanas** [comment]
> [@aplanas](https://github.com/aplanas) Latest btrfs-progs can create subvolumes without root and systemd-repart has been updated to make use of this.

Oh! Is this a compilation option?

```
# git version
> SYSTEMD_LOG_LEVEL=debug systemd-repart --empty=create --size=3G --definitions=./mkosi.repart ./test.img
Failed to find TPM2 pcrlock policy file 'pcrlock.json': No such file or directory
Loaded 'libcryptsetup.so.12' via dlopen()
Sized './test.img' to 3G.
Not resizing partition table, as there currently is none.
Sector size of device is 512 bytes. Using filesystem sector size of 4096 and grain size of 4096.
Pre-populating btrfs filesystem of partition 10-root.conf twice to calculate minimal partition size
Failed to make loopback device of /var/tmp/.#repart5a77754964eaafb9: Permission denied
```

```
> sudo btrfs --version
[sudo] password for root: 
btrfs-progs v6.14
-EXPERIMENTAL -INJECT -STATIC +LZO +ZSTD +UDEV +FSVERITY +ZONED CRYPTO=builtin
```

```
> systemd-repart --version
systemd 257 (257.5+suse.8.gc10a66fb4d)
+PAM +AUDIT +SELINUX +APPARMOR +IMA +IPE -SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBCRYPTSETUP_PLUGINS +LIBFDISK +PCRE2 +PWQUALITY +P11KIT +QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +BPF_FRAMEWORK -BTF -XKBCOMMON -UTMP +SYSVINIT +LIBARCHIVE
```

**Comment by aplanas** [comment]
Ah confirmed. It is working with systemd-repart 258 from git. Very nice!

**Comment by davidcassany** [comment]
Just wanted to add my two cents regarding this idea, which I think it is really interesting by the way. Feel free to ignore the comment, probably all this is already discussed, I just seen this issue now.

From a KIWI perspective, to my understanding, the challenge to is to create the root-tree that needs to be preloaded to disk at formatting time.

IMHO this requires a new logic in early create step to first build a temporary root-tree of each partition including all final artifacts (fstab, bootloader cfg, selinux attr, etc.) and subvolumes (e.g. the root-tree requires the full FS path `<root>/@/.snapshots/1/snapshot/<os-tree>`). With that logic in place it would be possible to build loopless images and eventually consider mount-less images in the future. IMHO having this logic is the real challenge, then wrapping systemd-repart or directly using the sgdisk and mkfs.* should be relatively straight forward (without systemd-repart preloading XFS will be painful though).

**Comment by keszybz** [comment]
> On RH/Fedora, it's part of the `systemd-udev` package (yeah, I dunno why, but oh well?).

The rule in Fedora is that parts of systemd to manage actual "hardware", i.e. in this case disks, virtual or not, are in `-udev` subpackage.
